File: solx/src/solx/output.py
"""Output layer: human Rich rendering vs machine-readable JSON.

Principle (issue #16 — "CLI design for agents"): a CLI driven by an agent
should not have to know a flag exists to get parseable output. So:

* When stdout is **not a TTY**, data commands emit JSON automatically; on a
  TTY they render Rich tables. The global `--json` flag forces JSON anywhere
  (a human on a terminal gets tables with no flag; the agent passes `--json`).
* All diagnostics, progress, and errors go to **stderr**, so stdout stays a
  clean data channel an agent can parse without stripping noise.
* Interactivity (whether we may *prompt*) is decided by **stdin**, separately
  from the stdout-format decision. A non-interactive session never blocks on
  a confirmation prompt (see `solx.jobs` / `solx.keep`).

`Out` bundles those three decisions plus the two streams so command bodies take
a single object and stay testable: a test builds an `Out` over ``StringIO``
streams with an explicit mode instead of poking globals.

**`rich` stays off the agent path.** On the JSON / non-interactive path
`Out.auto` builds a `_Plain` writer (plain text, markup stripped) instead of a
`rich.Console`, so an agent run (`--json`, or piped output) never imports
`rich` at all. `rich` is imported only when there's a human terminal to render
a table or coloured diagnostic for. Command modules import `rich.table` /
`rich.prompt` lazily for the same reason.
"""
from __future__ import annotations

import json as _json
import re
import sys
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Callable

if TYPE_CHECKING:
    from rich.console import Console


# Output mode override. The CLI sets "json" via the global --json flag; None
# means auto-detect from the stdout TTY. "plain" (force human) is supported by
# Out.auto for embedders/tests but has no CLI flag — a human on a terminal
# already gets human output by default, so forcing it isn't worth a flag.
Force = str  # "json" | "plain" | None


# Rich style tags ([red], [/], [bold dim], …). The char class deliberately
# excludes sentence punctuation (commas, quotes) so an interpolated exception
# string like "(at line 11, column 21)" isn't mistaken for markup. A literal
# bracket is written escaped as `\[` in our messages, so it is protected first.
_MARKUP = re.compile(r"\[/?[a-zA-Z0-9 #._]*\]")


def _plain(msg: str) -> str:
    """Strip Rich markup from `msg` for the no-Rich (agent/JSON) path."""
    msg = msg.replace("\\[", "\x00")  # protect escaped literal brackets
    msg = _MARKUP.sub("", msg)
    return msg.replace("\x00", "[")


class _Plain:
    """Minimal stand-in for `rich.Console` on the no-Rich path.

    Exposes just the slice command bodies (and tests) touch — `.print`, which
    strips markup and writes plain text, and `.file` — so nothing imports
    `rich` when output is JSON / agent-facing. Only ever receives diagnostic
    strings (the human table path constructs a real `rich.Console`).
    """

    is_terminal = False

    def __init__(self, file: Any) -> None:
        self.file = file

    def print(self, obj: Any = "") -> None:
        self.file.write(_plain(str(obj)) + "\n")


@dataclass
class Out:
    """A resolved output target: format choice + the two streams.

    * ``json_mode``  — emit JSON on the data channel (stdout) instead of Rich.
    * ``interactive`` — stdin is a TTY, so prompting a human is allowed.
    * ``stdout`` / ``stderr`` — the data and diagnostic writers: a
      ``rich.Console`` in human mode, a ``_Plain`` writer on the agent path.
      Both expose ``.print`` and ``.file``.
    """

    json_mode: bool
    interactive: bool
    stdout: "Console | _Plain"
    stderr: "Console | _Plain"

    @classmethod
    def auto(
        cls,
        *,
        force: Force | None = None,
        stdout: Any | None = None,
        stderr: Any | None = None,
        interactive: bool | None = None,
    ) -> "Out":
        """Build an `Out`, auto-detecting format from the stdout TTY.

        ``force`` (`"json"`/`"plain"`/`None`) overrides the auto-detect; the CLI
        passes `"json"` (global `--json`) or `None`. ``interactive`` defaults to
        whether **stdin** is a TTY. On the JSON path no `rich.Console` is built
        (and `rich` is never imported) — a `_Plain` writer is used instead.
        """
        # TTY-ness for format detection — from a caller-supplied stream/console
        # (tests, embedders) or sys.stdout (production), without importing rich.
        probe = stdout if stdout is not None else sys.stdout
        is_tty = getattr(probe, "is_terminal", None)
        if is_tty is None:
            try:
                is_tty = probe.isatty()
            except (AttributeError, ValueError, OSError):
                is_tty = False

        if force == "json":
            json_mode = True
        elif force == "plain":
            json_mode = False
        else:
            json_mode = not is_tty

        if interactive is None:
            try:
                interactive = sys.stdin.isatty()
            except (ValueError, OSError):
                interactive = False

        so, se = stdout, stderr
        if so is None or se is None:
            if json_mode:
                if so is None:
                    so = _Plain(sys.stdout)
                if se is None:
                    se = _Plain(sys.stderr)
            else:
                from rich.console import Console

                if so is None:
                    so = Console()
                if se is None:
                    se = Console(stderr=True)
        return cls(json_mode=json_mode, interactive=interactive, stdout=so, stderr=se)

    # --- diagnostics: always stderr, never on the JSON stdout stream --------

    def status(self, msg: str) -> None:
        """A progress / context line. Goes to stderr in every mode."""
        self.stderr.print(msg)

    def error(self, msg: str) -> None:
        """An error line. Goes to stderr in every mode."""
        self.stderr.print(msg)

    # --- data channel: stdout -----------------------------------------------

    def json(self, obj: Any) -> None:
        """Write one clean JSON document to stdout (no ANSI, no wrapping)."""
        # Write straight to the underlying file so Rich never injects color
        # or soft-wraps the payload, even under a forced `--json` on a TTY.
        self.stdout.file.write(_json.dumps(obj, indent=2, default=str) + "\n")
        self.stdout.file.flush()

    def human(self, renderable: Any) -> None:
        """Render something to stdout in human mode (Rich table, text, …)."""
        self.stdout.print(renderable)

    def emit(self, *, data: Any, human: Callable[[], Any]) -> None:
        """Emit a result: JSON ``data`` in json mode, else the ``human`` render.

        ``human`` is a thunk so the (possibly expensive) Rich renderable is
        only built when it will actually be shown.
        """
        if self.json_mode:
            self.json(data)
        else:
            rendered = human()
            if rendered is not None:
                self.stdout.print(rendered)

File: solx-rs/src/output.rs
//! Output layer: human rendering vs machine-readable JSON.
//!
//! A CLI driven by an agent should not have to know a flag exists to get
//! parseable output. So:
//!
//! * When stdout is **not a TTY**, data commands emit JSON automatically; on
//!   a TTY they render plain aligned tables. The global `--json` flag forces
//!   JSON anywhere (a human on a terminal gets tables with no flag; an agent
//!   passes `--json`).
//! * All diagnostics, progress, and errors go to **stderr**, so stdout stays
//!   a clean data channel an agent can parse without stripping noise.
//! * Interactivity (whether we may *prompt*) is decided by **stdin**,
//!   separately from the stdout-format decision. A non-interactive session
//!   never blocks on a confirmation prompt.
//!
//! JSON documents are rendered byte-for-byte like Python's
//! `json.dumps(obj, indent=2)` plus a trailing newline: two-space indent,
//! `", "` / `": "` separators, and `\uXXXX` escapes for every non-ASCII
//! character. Payloads are built as [`serde_json::Value`] with the
//! `preserve_order` feature, so object keys serialize in insertion order.

use std::io::{IsTerminal, Write};

use serde_json::Value;

/// A resolved output target: format choice + interactivity.
#[derive(Clone, Copy, Debug)]
pub struct Out {
    /// Emit JSON on the data channel (stdout) instead of a human rendering.
    pub json_mode: bool,
    /// stdin is a TTY, so prompting a human is allowed.
    pub interactive: bool,
}

impl Out {
    /// Build an `Out`, auto-detecting format from the stdout TTY.
    ///
    /// `force_json` (the global `--json` flag) overrides the auto-detect.
    /// `interactive` reflects whether **stdin** is a TTY.
    pub fn auto(force_json: bool) -> Self {
        Out {
            json_mode: force_json || !std::io::stdout().is_terminal(),
            interactive: std::io::stdin().is_terminal(),
        }
    }

    /// A progress / context line. Goes to stderr in every mode.
    pub fn status(&self, msg: &str) {
        eprintln!("{msg}");
    }

    /// An error line. Goes to stderr in every mode.
    pub fn error(&self, msg: &str) {
        eprintln!("{msg}");
    }

    /// Write one clean JSON document to stdout (no color, no wrapping).
    pub fn json(&self, obj: &Value) {
        let mut stdout = std::io::stdout().lock();
        let _ = stdout.write_all(to_python_json(obj).as_bytes());
        let _ = stdout.write_all(b"\n");
        let _ = stdout.flush();
    }

    /// Render something to stdout in human mode.
    pub fn human(&self, text: &str) {
        println!("{text}");
    }

    /// Emit a result: JSON `data` in json mode, else the `human` render.
    ///
    /// `human` is a thunk so the (possibly expensive) rendering is only
    /// built when it will actually be shown. A `None` render prints nothing.
    pub fn emit(&self, data: &Value, human: impl FnOnce() -> Option<String>) {
        if self.json_mode {
            self.json(data);
        } else if let Some(rendered) = human() {
            self.human(&rendered);
        }
    }
}

/// Ask a yes/no question on stderr and read the answer from stdin.
///
/// Callers gate on [`Out::interactive`] first — a non-interactive session
/// must never reach a prompt. Empty input takes `default`; `y`/`yes`
/// (case-insensitive) is true, anything else false.
pub fn confirm(prompt: &str, default: bool) -> bool {
    let hint = if default { "[Y/n]" } else { "[y/N]" };
    eprint!("{prompt} {hint} ");
    let _ = std::io::stderr().flush();
    let mut line = String::new();
    if std::io::stdin().read_line(&mut line).is_err() {
        return default;
    }
    let answer = line.trim().to_ascii_lowercase();
    if answer.is_empty() {
        return default;
    }
    matches!(answer.as_str(), "y" | "yes")
}

/// Render `v` exactly like Python's `json.dumps(v, indent=2)` (no trailing
/// newline; callers append one per document).
pub fn to_python_json(v: &Value) -> String {
    let mut buf = String::new();
    write_value(v, 0, &mut buf);
    buf
}

fn write_value(v: &Value, indent: usize, buf: &mut String) {
    match v {
        Value::Null => buf.push_str("null"),
        Value::Bool(b) => buf.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => buf.push_str(&n.to_string()),
        Value::String(s) => write_string(s, buf),
        Value::Array(items) => {
            if items.is_empty() {
                buf.push_str("[]");
                return;
            }
            buf.push_str("[\n");
            for (i, item) in items.iter().enumerate() {
                push_spaces(buf, indent + 2);
                write_value(item, indent + 2, buf);
                if i + 1 < items.len() {
                    buf.push(',');
                }
                buf.push('\n');
            }
            push_spaces(buf, indent);
            buf.push(']');
        }
        Value::Object(map) => {
            if map.is_empty() {
                buf.push_str("{}");
                return;
            }
            buf.push_str("{\n");
            for (i, (key, val)) in map.iter().enumerate() {
                push_spaces(buf, indent + 2);
                write_string(key, buf);
                buf.push_str(": ");
                write_value(val, indent + 2, buf);
                if i + 1 < map.len() {
                    buf.push(',');
                }
                buf.push('\n');
            }
            push_spaces(buf, indent);
            buf.push('}');
        }
    }
}

fn push_spaces(buf: &mut String, n: usize) {
    for _ in 0..n {
        buf.push(' ');
    }
}

/// Escape a string like Python's `json.dumps` with `ensure_ascii=True`:
/// everything outside `0x20..=0x7E` becomes a `\uXXXX` escape (surrogate
/// pairs for astral-plane characters).
fn write_string(s: &str, buf: &mut String) {
    buf.push('"');
    for c in s.chars() {
        match c {
            '"' => buf.push_str("\\\""),
            '\\' => buf.push_str("\\\\"),
            '\n' => buf.push_str("\\n"),
            '\r' => buf.push_str("\\r"),
            '\t' => buf.push_str("\\t"),
            '\u{8}' => buf.push_str("\\b"),
            '\u{c}' => buf.push_str("\\f"),
            '\u{20}'..='\u{7e}' => buf.push(c),
            _ => {
                let cp = c as u32;
                if cp <= 0xFFFF {
                    buf.push_str(&format!("\\u{cp:04x}"));
                } else {
                    let v = cp - 0x10000;
                    let hi = 0xD800 + (v >> 10);
                    let lo = 0xDC00 + (v & 0x3FF);
                    buf.push_str(&format!("\\u{hi:04x}\\u{lo:04x}"));
                }
            }
        }
    }
    buf.push('"');
}

/// Render `s` like Python's `repr()` for the common case: single quotes,
/// switching to double quotes when the string contains a single quote (and
/// no double quote), with backslash escapes for the usual control characters.
pub fn py_repr(s: &str) -> String {
    let quote = if s.contains('\'') && !s.contains('"') {
        '"'
    } else {
        '\''
    };
    let mut out = String::with_capacity(s.len() + 2);
    out.push(quote);
    for c in s.chars() {
        match c {
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if c == quote => {
                out.push('\\');
                out.push(c);
            }
            c if (c as u32) < 0x20 || c as u32 == 0x7f => {
                out.push_str(&format!("\\x{:02x}", c as u32));
            }
            c => out.push(c),
        }
    }
    out.push(quote);
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn json_object_two_space_indent_ordered() {
        let v = json!({"b": 1, "a": [1, 2], "c": {"x": true}});
        assert_eq!(
            to_python_json(&v),
            "{\n  \"b\": 1,\n  \"a\": [\n    1,\n    2\n  ],\n  \"c\": {\n    \"x\": true\n  }\n}"
        );
    }

    #[test]
    fn json_empty_containers_stay_inline() {
        assert_eq!(to_python_json(&json!([])), "[]");
        assert_eq!(to_python_json(&json!({})), "{}");
        assert_eq!(
            to_python_json(&json!({"a": [], "b": {}})),
            "{\n  \"a\": [],\n  \"b\": {}\n}"
        );
    }

    #[test]
    fn json_strings_escape_non_ascii_like_python() {
        // Python: json.dumps("café — ok\t\x7f") == '"caf\\u00e9 \\u2014 ok\\t\\u007f"'
        let v = json!("café — ok\t\u{7f}");
        assert_eq!(to_python_json(&v), "\"caf\\u00e9 \\u2014 ok\\t\\u007f\"");
    }

    #[test]
    fn json_astral_plane_uses_surrogate_pairs() {
        let v = json!("🎉");
        assert_eq!(to_python_json(&v), "\"\\ud83c\\udf89\"");
    }

    #[test]
    fn json_null_and_numbers() {
        let v = json!({"keep": null, "n": 300});
        assert_eq!(to_python_json(&v), "{\n  \"keep\": null,\n  \"n\": 300\n}");
    }

    #[test]
    fn py_repr_quoting() {
        assert_eq!(py_repr("tcsh"), "'tcsh'");
        assert_eq!(py_repr("it's"), "\"it's\"");
        assert_eq!(py_repr("a\"b'c"), "'a\"b\\'c'");
        assert_eq!(py_repr("a\nb"), "'a\\nb'");
    }
}
